Add threadpool

BitSyndicate1 opened the pull request:
setup global threadpool

anonymous123-code commented on vr-threading/src/lib.rs:
```diff
@@ -1,14 +1,107 @@
+
+pub mod global_threadpool;
+
+pub static mut THREADPOOL: StandardThreadPool = StandardThreadPool::new();
```
I don't think that the thread pool should be global.

anonymous123-code commented on vr-threading/src/lib.rs:
```diff
@@ -1,14 +1,107 @@
+            workers: Vec::new(),
+        }
+    }
+    unsafe fn init(&mut self, size: NonZeroUsize) {
```
Is this really unsafe?

anonymous123-code commented on vr-threading/src/lib.rs:
```diff
@@ -1,14 +1,107 @@
+            }
+        }
     }
 }
```
Test pls

anonymous123-code commented on vr-threading/src/lib.rs:
```diff
@@ -1,14 +1,107 @@
+        for worker in self.workers.iter_mut() {
+            println!("ThreadPool: Shutting down worker {}", worker.id);
+            if let Some(thread) = worker.thread.take() {
+                thread.join().unwrap();
```
If join waits for all pools to finish, but a pool has no task, this might hold, I think.

BitSyndicate1 replied:
No it is fully valid.

BitSyndicate1 commented:
Not finished working out memory freeing problems

BitSyndicate1 committed:
add support for mach_o and coff systems for the threadpool termination

BitSyndicate1 committed:
make global threadpool private for encapsulation

BitSyndicate1 committed:
Add threadpool
setup global threadpool

BitSyndicate1 committed:
fix formatting

BitSyndicate1 committed:
Use lazy static for global threadpool
less overhead and only a macro crate

BitSyndicate1 committed:
fix formatting

BitSyndicate1 committed:
finish

BitSyndicate1 committed:
Fix cleanup problem with global threadpool
Uses the fini_array sector of the binary to cleanup the static
THREADPOOL. initializes with global_init, in future this will be moved

BitSyndicate1 committed:
make destructor right

BitSyndicate1 committed:
Remove unnecessary dependency, format right

BitSyndicate1 committed:
fix main put global threadpool init top

BitSyndicate1 committed:
Use logger for threadpool messages

BitSyndicate1 committed:
logger destructor after tp fix some stuff

anonymous123-code committed:
Improve formatting API for logger

anonymous123-code committed:
Better thread pool error handling

anonymous123-code committed:
ThreadPool: Add global_eval

BitSyndicate1 committed:
Fix gray formatting in logger and add promise API

BitSyndicate1 committed:
some minor adjustments

BitSyndicate1 committed:
Document the vr-threading crate

BitSyndicate1 committed:
format stuff

BitSyndicate1 committed:
use a global threadpool handle for the shutdown to be in main, removing binary hacks
